node/node.go

```
@@ -130,6 +133,82 @@ func StateProvider(stateProvider statesync.StateProvider) Option {
+	ctx := context.Background()
+	logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+	if config == nil {
+		config = cfg.DefaultConfig()
+	}
+	blockStore, stateDB, err := initDBs(config, cfg.DefaultDBProvider)
+
+	if err != nil {
+		return err
+	}
+
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: config.Storage.DiscardABCIResponses,
+	})
+
+	if !blockStore.IsEmpty() {
+		logger.Error("overwriting non empty block store")
+	}
+	genState, _, err := LoadStateFromDBOrGenesisDocProvider(stateDB, DefaultGenesisDocProviderFunc(config))
+	if err != nil {
+		return err
+	}
+
+	stateProvider, err := statesync.NewLightClientStateProvider(
+		ctx,
+		genState.ChainID, genState.Version, genState.InitialHeight,
+		config.StateSync.RPCServers, light.TrustOptions{
+			Period: config.StateSync.TrustPeriod,
+			Height: config.StateSync.TrustHeight,
+			Hash:   config.StateSync.TrustHashBytes(),
+		}, logger.With("module", "light"))
+	if err != nil {
+		return fmt.Errorf("failed to set up light client state provider: %w", err)
+	}
+
+	state, err := stateProvider.State(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	commit, err := stateProvider.Commit(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	if err := stateStore.Bootstrap(state); err != nil {
+		return err
+	}
+
+	// store.SaveBlockStoreState(&cmtstore.BlockStoreState{
+	// 	// it breaks the invariant that blocks in range [Base, Height] must exists, but it do works in practice.
+
+	// 	Base:   state.LastBlockHeight,
+	// 	Height: state.LastBlockHeight,
+	// }, blockStore)
+	err = blockStore.SaveSeenCommit(state.LastBlockHeight, commit)
+	if err != nil {
+		return err
+	}
+
+	if appHash == nil {
+		logger.Info("warning: cannot verify appHash. Verification will happen when node boots up!")
+	} else {
+		if !bytes.Equal(appHash, state.AppHash) {
+			logger.Error("the app hash returned by the light client does not match the provided appHash, expected %X, got %X", state.AppHash, appHash)
```

I would return an error in this case